### PR TITLE
feat(AllocatedBuffer): allow buffer to have void type

### PR DESCRIPTION
**Pull request by zcorniere**
In the future Transient Ressource System, I will need to store multiple templated version of the `AllocatedBuffer`.

This PR allow a void type buffer, **that is still valid to use**, but this allow to store AllocatedBuffer into a `std::vector<AllocatedBuffer<void>>` without having issue with the Template parameters.

Also, converting between `AllocatedBuffer<void>` to `AllocatedBuffer<T>` is allowed, but not the other way around.

### CODE REVIEW

**Comment by Ersikan** [comment]
Why not `AllocatedBuffer<std::byte>` as a default ?

**Comment by zcorniere** [comment]
I thought of this too, but I think `void` carry more meaning toward "anything can be in there/we don't care what could be in there" than `std::byte`.

`AllocatedBuffer<std::byte>` is a true and valid use case of an AllocatedBuffer, whereas `AllocatedBuffer<void>` hint that we don't care what is in it and we shouldn't try to access it.

File: libgraphics/include/pivot/graphics/types/AllocatedBuffer.hxx [hunk 1]
```diff
@@ -8,26 +8,41 @@ namespace pivot::graphics
...
+
+    /// @brief get the byte size
+    ///
+    /// When `T` is `void`, the size of buffer will be multiplied by `sizeof(T)`
+    auto getBytesSize() const noexcept requires(!std::is_same_v<T, void>) { return getSize() * sizeof(T); }
+    /// @copybrief getBytesSize
+    ///
+    /// When `T` is `void`, the size of buffer will be returned; as void does not have a size, we assume it is a buffer
+    /// of bytes
+    auto getBytesSize() const noexcept requires(std::is_same_v<T, void>) { return getSize(); }
```
Comment (Ersikan) [inline libgraphics/include/pivot/graphics/types/AllocatedBuffer.hxx:45]: Use `if constexpr` to create only one method:
```suggestion
    auto getBytesSize() const noexcept requires(!std::is_same_v<T, void>) {
        if constexpr (std::is_same_v<T, void>) {
            return getSize();
        } else {
            return getSize() * sizeof(T);
        }
     }
```

File: libgraphics/include/pivot/graphics/types/AllocatedBuffer.hxx [hunk 2]
```diff
@@ -58,6 +73,20 @@ public:
...
+    operator AllocatedBuffer<O>() const noexcept { return *this; }
+
+    template <BufferValid O>
+    /// Allow to convert AllocatedBuffer<void> to AllocatedBuffer<T>
+    requires std::is_same_v<T, void> AllocatedBuffer<T>
+    operator=(const AllocatedBuffer<O> &other) const noexcept
+    {
+        *this = other;
+        return *this;
+    }
```
Comment (Ersikan) [inline libgraphics/include/pivot/graphics/types/AllocatedBuffer.hxx:88]: Maybe do some checks that the conversion is valid or something ?

**Review by Ersikan — CHANGES_REQUESTED** [review]

**Review by Ersikan — APPROVED** [review]

**Pull request merged**